projects/gloo/cli/pkg/snapshot/domain.go

```diff
@@ -0,0 +1,348 @@
+		i.httpRoutes = make(map[string]*HTTPRouteWrapper)
+	}
+	i.httpRoutes[route.Index()] = route
+}
+
+func (i *Instance) AddListenerSet(l *ListenerSetWrapper) {
+	if i.listenerSets == nil {
+		i.listenerSets = make(map[string]*ListenerSetWrapper)
+	}
+	i.listenerSets[l.Index()] = l
+}
+
+func (i *Instance) AddDirectResponse(d *DirectResponseWrapper) {
+	if i.directResponses == nil {
+		i.directResponses = make(map[string]*DirectResponseWrapper)
+	}
+	i.directResponses[d.Index()] = d
+}
+
+func (i *Instance) AddSettings(d *SettingsWrapper) {
+	if i.settings == nil {
+		i.settings = make(map[string]*SettingsWrapper)
+	}
+	i.settings[d.Index()] = d
+}
+
+func (i *Instance) AddAuthConfig(d *AuthConfigWrapper) {
+
+	if i.authConfigs == nil {
+		i.authConfigs = make(map[string]*AuthConfigWrapper)
+	}
+	i.authConfigs[d.Index()] = d
+}
+
+func (i *Instance) AddUpstream(u *UpstreamWrapper) {
+	if i.upstreams == nil {
+		i.upstreams = make(map[string]*UpstreamWrapper)
+	}
+	i.upstreams[u.Index()] = u
+}
+
+func (i *Instance) AddRouteTable(r *RouteTableWrapper) {
+	if i.routeTables == nil {
+		i.routeTables = make(map[string]*RouteTableWrapper)
+	}
+	i.routeTables[r.Index()] = r
+}
+
+func (i *Instance) AddVirtualService(v *VirtualServiceWrapper) {
+	if i.virtualServices == nil {
+		i.virtualServices = make(map[string]*VirtualServiceWrapper)
+	}
+	i.virtualServices[v.Index()] = v
+}
+
+func (i *Instance) AddRouteOption(r *RouteOptionWrapper) {
+	if i.routeOptions == nil {
+		i.routeOptions = make(map[string]*RouteOptionWrapper)
+	}
+	i.routeOptions[r.Index()] = r
+}
+
+func (i *Instance) AddVirtualHostOption(v *VirtualHostOptionWrapper) {
+	if i.virtualHostOptions == nil {
+		i.virtualHostOptions = make(map[string]*VirtualHostOptionWrapper)
+	}
+	i.virtualHostOptions[v.Index()] = v
+}
+
+func (i *Instance) AddGlooGateway(g *GlooGatewayWrapper) {
+	if i.glooGateways == nil {
+		i.glooGateways = make(map[string]*GlooGatewayWrapper)
+	}
+	i.glooGateways[g.Index()] = g
+}
+
+func (i *Instance) AddHTTPListenerOption(h *HTTPListenerOptionWrapper) {
+	if i.httpListenerOptions == nil {
+		i.httpListenerOptions = make(map[string]*HTTPListenerOptionWrapper)
+	}
+	i.httpListenerOptions[h.Index()] = h
+}
+
+func (i *Instance) AddListenerOption(l *ListenerOptionWrapper) {
+	if i.listenerOptions == nil {
+		i.listenerOptions = make(map[string]*ListenerOptionWrapper)
+	}
+	i.listenerOptions[l.Index()] = l
+}
+
+func (i *Instance) AddYamlObject(y *YAMLWrapper) {
+	if i.yamlObjects == nil {
+		i.yamlObjects = []*YAMLWrapper{}
+	}
+	i.yamlObjects = append(i.yamlObjects, y)
+}
+
+// NameNamespaceIndex is the unique identifier that identifies a specific one
+func NameNamespaceIndex(name string, namespace string) string {
+	return fmt.Sprintf("%s/%s", namespace, name)
+}
+
+// GlooGatewayVirtualServices finds all VirtualServiceWrapper that match the gateways selectors
+func (i *Instance) GlooGatewayVirtualServices(gateway *GlooGatewayWrapper) ([]*VirtualServiceWrapper, error) {
```

Comment:
This method, and the ones below, seem like they define logic/behavior that is ingrained in the product, and we are trying to mirror it. My concern with redefining code in the CLI, is that it is likely to drift over time from the source implementation. Is there code in the product that we could import/invoke here instead? If I'm off, and this isn't true functionality, then just let me know, and I'm good with what you have

Reply:
The problem is that snapshots are not well formed with the original golang objects. Ideally the snapshot format would be more like how its stored in code but we need to precondition the objects to have the correct fields to make the selectors work

I still use the translator but i have to prep the objects beforehand
```golang
		matches, err := translator.HttpGatewayContainsVirtualService(gateway.Gateway.Spec.GetHttpGateway(), vs, gateway.Gateway.Spec.Ssl)
```